File: LUSIA Studio - Backend/app/api/http/routers/worksheet_generation.py
"""
Worksheet generation endpoints — AI-powered worksheet creation pipeline.

Provides endpoints for:
- POST /start: create artifact, assemble context, generate initial blueprint
- GET /{artifact_id}/blueprint: fetch current blueprint state (session recovery)
- POST /{artifact_id}/blueprint/chat: process one chat turn during review
- PATCH /{artifact_id}/blueprint: direct UI edits (drag-reorder, etc.)
- GET /{artifact_id}/resolve/stream: SSE stream for resolution progress
"""

from fastapi import APIRouter, Depends, HTTPException
from fastapi.responses import StreamingResponse
from supabase import Client

from app.api.deps import require_teacher
from app.api.http.schemas.worksheet_generation import (
    BlueprintChatIn,
    BlueprintStateOut,
    BlueprintUpdateIn,
    ContextSummary,
    TemplateInfo,
    TemplateListOut,
    WorksheetStartIn,
    WorksheetStartOut,
)
from app.api.http.services.worksheet_blueprint_agent import (
    stream_blueprint_chat_turn,
)
from app.api.http.services.generation_context import (
    fetch_bank_questions,
    validate_generation_possible,
)
from app.api.http.services.worksheet_generation_service import (
    create_worksheet_artifact,
    get_worksheet_artifact,
    update_worksheet_content,
)
from app.api.http.services.worksheet_planner import generate_initial_blueprint, stream_initial_blueprint
from app.api.http.services.worksheet_resolution import resolve_worksheet_stream
from app.core.database import get_b2b_db

router = APIRouter()


@router.get("/templates", response_model=TemplateListOut)
async def get_templates():
    """Return all available worksheet templates."""
    from app.api.http.services.worksheet_templates import (
        get_template_info,
        list_templates,
    )

    templates = list_templates()
    return TemplateListOut(
        templates=[TemplateInfo(**get_template_info(t)) for t in templates]
    )


@router.post("/start", response_model=WorksheetStartOut, status_code=201)
async def start_worksheet_generation(
    payload: WorksheetStartIn,
    current_user: dict = Depends(require_teacher),
    db: Client = Depends(get_b2b_db),
):
    """
    Create a worksheet artifact and return its ID immediately.

    Blueprint generation is triggered by the client connecting to
    GET /{artifact_id}/blueprint/stream right after navigation.
    """
    can_proceed, reason = validate_generation_possible(
        db, payload.subject_id, payload.upload_artifact_id
    )
    if not can_proceed:
        raise HTTPException(status_code=400, detail=reason)

    artifact = create_worksheet_artifact(
        db,
        current_user["organization_id"],
        current_user["id"],
        payload,
    )

    return WorksheetStartOut(
        artifact_id=artifact["id"],
        artifact_name=artifact["artifact_name"],
    )


@router.get("/{artifact_id}/blueprint", response_model=BlueprintStateOut)
async def get_blueprint(
    artifact_id: str,
    current_user: dict = Depends(require_teacher),
    db: Client = Depends(get_b2b_db),
):
    """
    Fetch the current blueprint state, conversation history, and context
    summary. Used for page load and session recovery.
    """
    artifact = get_worksheet_artifact(db, artifact_id, current_user["id"])
    content = artifact.get("content", {})

    from app.api.http.schemas.worksheet_generation import Blueprint

    raw_blueprint = content.get("blueprint", {})
    blueprint = Blueprint(**(raw_blueprint or {}))
    conversation = content.get("conversation_history", [])
    generation_params = content.get("generation_params", {})

    # Build context summary
    summary_data = content.get("assembled_context_summary", {})
    context_summary = ContextSummary(
        subject_name=summary_data.get("subject_name", ""),
        subject_status=summary_data.get("subject_status", ""),
        has_national_exam=summary_data.get("has_national_exam", False),
        bank_question_count=summary_data.get("bank_question_count", 0),
        document_attached=summary_data.get("document_attached", False),
        curriculum_code_count=summary_data.get("curriculum_code_count", 0),
    )

    return BlueprintStateOut(
        blueprint=blueprint,
        conversation=conversation,
        generation_params=generation_params,
        context_summary=context_summary,
    )


@router.get("/{artifact_id}/blueprint/stream")
async def stream_blueprint_generation(
    artifact_id: str,
    current_user: dict = Depends(require_teacher),
    db: Client = Depends(get_b2b_db),
):
    """
    SSE stream for initial blueprint generation.

    Assembles context and runs the planner LLM via instructor's create_iterable,
    emitting each BlueprintBlock as it completes. If the blueprint was already
    generated, replays existing blocks immediately.

    Events: {"type": "block", "block": {...}} | {"type": "done"} | {"type": "error", "message": "..."}
    """
    generator = stream_initial_blueprint(
        db,
        artifact_id,
        current_user["organization_id"],
        current_user["id"],
    )
    return StreamingResponse(
        generator,
        media_type="text/event-stream",
        headers={
            "Cache-Control": "no-cache",
            "Connection": "keep-alive",
            "X-Accel-Buffering": "no",
        },
    )


@router.post("/{artifact_id}/blueprint/chat")
async def blueprint_chat(
    artifact_id: str,
    payload: BlueprintChatIn,
    current_user: dict = Depends(require_teacher),
    db: Client = Depends(get_b2b_db),
):
    """
    SSE stream for one teacher chat turn during blueprint review.

    Streams upsert/delete events as each tool call is applied, then
    a final 'done' event with the full blueprint. Persists state after
    the stream completes.
    """
    artifact = get_worksheet_artifact(db, artifact_id, current_user["id"])
    content = artifact.get("content", {})
    conversation = content.get("conversation_history", [])

    # Fetch bank questions for the agent context
    subject_id = artifact["subject_id"]
    year_level = artifact["year_level"]
    curriculum_codes = artifact.get("curriculum_codes") or []

    bank_questions = fetch_bank_questions(
        db, subject_id, year_level, curriculum_codes
    )

    async def generate():
        import json as _json

        all_tool_calls = []
        final_blueprint = None

        async for chunk in stream_blueprint_chat_turn(
            message=payload.message,
            block_id=payload.block_id,
            current_blueprint=payload.blueprint,
            conversation_history=conversation,
            bank_questions=bank_questions,
        ):
            yield chunk

            # Parse the done event to persist state
            stripped = chunk.strip()
            if stripped.startswith("data: "):
                try:
                    event = _json.loads(stripped[6:])
                    if event.get("type") == "done":
                        final_blueprint = event.get("blueprint")
                        all_tool_calls = event.get("tool_calls", [])
                except _json.JSONDecodeError:
                    pass

        # Persist after stream completes
        if final_blueprint is not None:
            from app.api.http.schemas.worksheet_generation import Blueprint

            bp = Blueprint(**final_blueprint)

            user_entry = {"role": "user", "content": payload.message}
            if payload.block_id:
                user_entry["block_id"] = payload.block_id

            conversation.append(user_entry)
            conversation.append({
                "role": "assistant",
                "content": "",
                "tool_calls": all_tool_calls,
            })

            content["blueprint"] = bp.model_dump()
            content["conversation_history"] = conversation
            content["phase"] = "blueprint_review"
            update_worksheet_content(db, artifact_id, content)

    return StreamingResponse(
        generate(),
        media_type="text/event-stream",
        headers={
            "Cache-Control": "no-cache",
            "Connection": "keep-alive",
            "X-Accel-Buffering": "no",
        },
    )


@router.patch("/{artifact_id}/blueprint")
async def update_blueprint_direct(
    artifact_id: str,
    payload: BlueprintUpdateIn,
    current_user: dict = Depends(require_teacher),
    db: Client = Depends(get_b2b_db),
):
    """
    Direct blueprint update from the UI (drag-reorder, manual block edits).

    Called by the frontend on a debounce. Does not involve AI.
    """
    artifact = get_worksheet_artifact(db, artifact_id, current_user["id"])
    content = artifact.get("content", {})

    content["blueprint"] = payload.blueprint.model_dump()
    update_worksheet_content(db, artifact_id, content)

    return {"status": "ok"}


@router.get("/{artifact_id}/resolve/stream")
async def stream_resolution(
    artifact_id: str,
    current_user: dict = Depends(require_teacher),
    db: Client = Depends(get_b2b_db),
):
    """
    SSE endpoint for worksheet resolution.

    Streams events as bank blocks are fetched and AI blocks are generated
    in parallel. The frontend receives events for each resolved question.
    """
    generator = resolve_worksheet_stream(
        db,
        artifact_id,
        current_user["organization_id"],
        current_user["id"],
    )
    return StreamingResponse(
        generator,
        media_type="text/event-stream",
        headers={
            "Cache-Control": "no-cache",
            "Connection": "keep-alive",
            "X-Accel-Buffering": "no",
        },
    )

File: LUSIA Studio - Frontend/lib/worksheet-generation.ts
/**
 * Worksheet Generation — TypeScript types & API client
 *
 * Handles the worksheet generation endpoints:
 * - POST /api/worksheet-generation/start
 * - GET /api/worksheet-generation/{artifactId}/blueprint
 * - POST /api/worksheet-generation/{artifactId}/blueprint/chat
 * - PATCH /api/worksheet-generation/{artifactId}/blueprint
 * - GET /api/worksheet-generation/{artifactId}/resolve/stream (SSE)
 */

/* ═══════════════════════════════════════════════════════════════
   TYPES
   ═══════════════════════════════════════════════════════════════ */

export interface WorksheetStartInput {
    subject_id?: string | null;
    year_level?: string | null;
    subject_component?: string | null;
    curriculum_codes: string[];
    upload_artifact_id?: string | null;
    prompt: string;
    template_id: string;
    difficulty: "Fácil" | "Médio" | "Difícil";
    year_range?: [number, number] | null;
}

export interface TemplateInfo {
    id: string;
    name: string;
    tier: "quick" | "practice" | "exam";
    description: string;
    estimated_minutes: string;
    group_count: number;
    total_slots: number;
}

export interface WorksheetStartResult {
    artifact_id: string;
    artifact_name: string;
    artifact_type: string;
    icon: string | null;
    source_type: string;
    subject_id: string | null;
    subject_ids: string[] | null;
    year_level: string | null;
    curriculum_codes: string[] | null;
    is_processed: boolean;
    is_public: boolean;
    created_at: string | null;
}

export interface BlueprintBlock {
    id: string;
    order: number;
    source: "bank" | "ai_generated";
    question_id?: string | null;
    curriculum_code: string;
    curriculum_path?: string | null;
    type: string;
    goal: string;
    difficulty?: string | null;
    group_label?: string | null;
    reference_question_ids: string[];
    children?: BlueprintBlock[] | null;
}

export interface Blueprint {
    blocks: BlueprintBlock[];
    version: number;
}

export interface ContextSummary {
    subject_name: string;
    subject_status: string;
    has_national_exam: boolean;
    bank_question_count: number;
    document_attached: boolean;
    curriculum_code_count: number;
}

export interface BlueprintState {
    blueprint: Blueprint;
    conversation: ChatMessage[];
    generation_params: Record<string, any>;
    context_summary: ContextSummary;
}

export interface ChatMessage {
    role: "user" | "assistant";
    content: string;
    block_id?: string | null;
    tool_calls?: ToolCallRecord[];
}

export interface ToolCallRecord {
    name: string;
    args: Record<string, any>;
    result?: {
        action: string;
        affected_block_ids: string[];
        message: string;
        parent_id?: string | null;
        block?: BlueprintBlock;
    };
}

export interface BlueprintChatResponse {
    message: string;
    blueprint: Blueprint;
    tool_calls: ToolCallRecord[];
}

export type BlueprintStreamEvent =
    | { type: "block"; block: BlueprintBlock }
    | { type: "child_block"; parent_id: string; block: BlueprintBlock }
    | { type: "done" }
    | { type: "error"; message: string };

export type BlueprintChatStreamEvent =
    | { type: "mutation"; mutation: { action: string; affected_block_ids: string[]; message: string; parent_id?: string | null; block?: BlueprintBlock } }
    | { type: "done"; blueprint: Blueprint; tool_calls: ToolCallRecord[] }
    | { type: "error"; message: string };

export type WorksheetStreamEvent =
    | { type: "started"; total_blocks: number }
    | { type: "bank_resolved"; block_id: string; question_id: string; question_type: string; order: number; top_level_order: number; child_order?: number | null; question_content?: Record<string, any>; parent_question_id?: string; parent_block_id?: string | null }
    | { type: "question"; block_id: string; question_id: string; question_type: string; label: string; order: number; top_level_order: number; child_order?: number | null; question_content?: Record<string, any>; parent_question_id?: string; parent_block_id?: string | null }
    | { type: "block_error"; block_id: string; message: string }
    | { type: "block_warning"; block_id: string; message: string }
    | { type: "done"; artifact_id: string; total_questions: number }
    | { type: "error"; message: string };


/* ═══════════════════════════════════════════════════════════════
   API FUNCTIONS
   ═══════════════════════════════════════════════════════════════ */

/**
 * Create a worksheet artifact and generate the initial blueprint.
 */
export async function startWorksheetGeneration(
    input: WorksheetStartInput,
): Promise<WorksheetStartResult> {
    const response = await fetch("/api/worksheet-generation/start", {
        method: "POST",
        headers: { "Content-Type": "application/json" },
        body: JSON.stringify(input),
        cache: "no-store",
    });

    if (!response.ok) {
        const err = await response.json().catch(() => ({}));
        throw new Error(err.detail || err.error || "Failed to start worksheet generation");
    }

    const result = await response.json();
    const { cacheInvalidate } = await import("@/lib/cache");
    cacheInvalidate("artifacts:");
    return result;
}

/**
 * Fetch the current blueprint state (for page load / session recovery).
 */
export async function getBlueprintState(
    artifactId: string,
): Promise<BlueprintState> {
    const response = await fetch(
        `/api/worksheet-generation/${artifactId}/blueprint`,
        { cache: "no-store" },
    );

    if (!response.ok) {
        const err = await response.json().catch(() => ({}));
        throw new Error(err.detail || err.error || "Failed to fetch blueprint");
    }

    return response.json();
}

/**
 * Send a chat message during the blueprint review phase.
 */
export async function sendBlueprintChat(
    artifactId: string,
    message: string,
    blueprint: Blueprint,
    blockId?: string | null,
): Promise<BlueprintChatResponse> {
    const response = await fetch(
        `/api/worksheet-generation/${artifactId}/blueprint/chat`,
        {
            method: "POST",
            headers: { "Content-Type": "application/json" },
            body: JSON.stringify({
                message,
                block_id: blockId || null,
                blueprint,
            }),
            cache: "no-store",
        },
    );

    if (!response.ok) {
        const err = await response.json().catch(() => ({}));
        throw new Error(err.detail || err.error || "Failed to process chat message");
    }

    return response.json();
}

/**
 * Open an SSE stream for a blueprint chat turn.
 *
 * Sends a POST with the message + current blueprint, then streams
 * upsert/delete events as the AI applies tool calls. Returns an
 * AbortController for cleanup.
 */
export function streamBlueprintChat(
    artifactId: string,
    message: string,
    blueprint: Blueprint,
    blockId: string | null,
    onEvent: (event: BlueprintChatStreamEvent) => void,
    onError: (error: Error) => void,
): AbortController {
    const controller = new AbortController();

    (async () => {
        try {
            const response = await fetch(
                `/api/worksheet-generation/${artifactId}/blueprint/chat`,
                {
                    method: "POST",
                    headers: {
                        "Content-Type": "application/json",
                        Accept: "text/event-stream",
                    },
                    body: JSON.stringify({
                        message,
                        block_id: blockId || null,
                        blueprint,
                    }),
                    signal: controller.signal,
                    cache: "no-store",
                },
            );

            if (!response.ok) {
                throw new Error(`Chat stream failed with status ${response.status}`);
            }

            const reader = response.body?.getReader();
            if (!reader) throw new Error("No response body");

            const decoder = new TextDecoder();
            let buffer = "";

            while (true) {
                const { done, value } = await reader.read();
                if (done) break;

                buffer += decoder.decode(value, { stream: true });
                const lines = buffer.split("\n");
                buffer = lines.pop() || "";

                for (const line of lines) {
                    const trimmed = line.trim();
                    if (!trimmed || trimmed.startsWith(":")) continue;
                    if (!trimmed.startsWith("data: ")) continue;

                    try {
                        const event = JSON.parse(trimmed.slice(6)) as BlueprintChatStreamEvent;
                        onEvent(event);
                    } catch {
                        // Skip malformed JSON
                    }
                }
            }
        } catch (err) {
            if (controller.signal.aborted) return;
            onError(err instanceof Error ? err : new Error(String(err)));
        }
    })();

    return controller;
}

/**
 * Save direct UI edits to the blueprint (drag-reorder, etc.).
 * Call this on a debounce from the frontend.
 */
export async function updateBlueprint(
    artifactId: string,
    blueprint: Blueprint,
): Promise<void> {
    const response = await fetch(
        `/api/worksheet-generation/${artifactId}/blueprint`,
        {
            method: "PATCH",
            headers: { "Content-Type": "application/json" },
            body: JSON.stringify({ blueprint }),
            cache: "no-store",
        },
    );

    if (!response.ok) {
        const err = await response.json().catch(() => ({}));
        throw new Error(err.detail || err.error || "Failed to update blueprint");
    }
}

/**
 * Open an SSE stream for real-time blueprint generation.
 *
 * Emits BlueprintBlock objects one by one as instructor parses them from the
 * streaming LLM response. Call immediately after navigating to the blueprint page.
 */
export function streamBlueprintGeneration(
    artifactId: string,
    onEvent: (event: BlueprintStreamEvent) => void,
    onError: (error: Error) => void,
    onComplete: () => void,
): AbortController {
    const controller = new AbortController();

    (async () => {
        try {
            const response = await fetch(
                `/api/worksheet-generation/${artifactId}/blueprint/stream`,
                {
                    method: "GET",
                    headers: { Accept: "text/event-stream" },
                    signal: controller.signal,
                    cache: "no-store",
                },
            );

            if (!response.ok) {
                throw new Error(`Stream request failed with status ${response.status}`);
            }

            const reader = response.body?.getReader();
            if (!reader) throw new Error("No response body");

            const decoder = new TextDecoder();
            let buffer = "";

            while (true) {
                const { done, value } = await reader.read();
                if (done) break;

                buffer += decoder.decode(value, { stream: true });
                const lines = buffer.split("\n");
                buffer = lines.pop() || "";

                for (const line of lines) {
                    const trimmed = line.trim();
                    if (!trimmed || trimmed.startsWith(":")) continue;
                    if (!trimmed.startsWith("data: ")) continue;

                    try {
                        const event = JSON.parse(trimmed.slice(6)) as BlueprintStreamEvent;
                        onEvent(event);
                        if (event.type === "done" || event.type === "error") {
                            onComplete();
                            return;
                        }
                    } catch {
                        // Skip malformed JSON
                    }
                }
            }

            onComplete();
        } catch (err) {
            if (controller.signal.aborted) return;
            onError(err instanceof Error ? err : new Error(String(err)));
        }
    })();

    return controller;
}


/**
 * Fetch all available worksheet templates.
 */
export async function getWorksheetTemplates(): Promise<TemplateInfo[]> {
    const response = await fetch("/api/worksheet-generation/templates", {
        cache: "no-store",
    });

    if (!response.ok) {
        throw new Error("Failed to fetch templates");
    }

    const data = await response.json();
    return data.templates;
}

/**
 * Open an SSE stream for worksheet resolution.
 *
 * Returns an AbortController for cleanup on component unmount.
 */
export function streamWorksheetResolution(
    artifactId: string,
    onEvent: (event: WorksheetStreamEvent) => void,
    onError: (error: Error) => void,
    onComplete: () => void,
): AbortController {
    const controller = new AbortController();

    (async () => {
        try {
            const response = await fetch(
                `/api/worksheet-generation/${artifactId}/resolve/stream`,
                {
                    method: "GET",
                    headers: { Accept: "text/event-stream" },
                    signal: controller.signal,
                    cache: "no-store",
                },
            );

            if (!response.ok) {
                throw new Error(`Stream request failed with status ${response.status}`);
            }

            const reader = response.body?.getReader();
            if (!reader) {
                throw new Error("No response body");
            }

            const decoder = new TextDecoder();
            let buffer = "";

            while (true) {
                const { done, value } = await reader.read();
                if (done) break;

                buffer += decoder.decode(value, { stream: true });

                const lines = buffer.split("\n");
                buffer = lines.pop() || "";

                for (const line of lines) {
                    const trimmed = line.trim();
                    if (!trimmed || trimmed.startsWith(":")) continue;

                    if (trimmed.startsWith("data: ")) {
                        const jsonStr = trimmed.slice(6);
                        try {
                            const event = JSON.parse(jsonStr) as WorksheetStreamEvent;
                            onEvent(event);

                            if (event.type === "done" || event.type === "error") {
                                onComplete();
                                return;
                            }
                        } catch {
                            // Skip malformed JSON
                        }
                    }
                }
            }

            onComplete();
        } catch (err) {
            if (controller.signal.aborted) return;
            onError(err instanceof Error ? err : new Error(String(err)));
        }
    })();

    return controller;
}
